analysis/metrics.py

```diff
@@ -143,3 +144,157 @@ class AggregateMetrics:
+    #
+    # Consider the following example with a single partition to see how they are
+    # different from abs/rel_error_expected metrics:
+    #
+    # Given 1 partition with probability_to_keep=0.4, actual_count=100,
+    # abs_error_expected=-50;
+    # -> abs_error_expected = (0.4*-50)/0.4=-50
+    # -> abs_error_expected_w_dropped_partitions = 0.4*-50+0.6*-100=-80
+    #
+    # When public partitions are used, these will be exactly equal to
+    # abs/rel_error_expected.
+    with_dropped_partitions: float
+
+
+@dataclass
+class DroppedDataInfo:
+    """Information about the data dropped during different procedures.
+
+    Attributes:
+        l0: ratio of data dropped during of l0 contribution bounding.
+        linf: ratio of data dropped during of linf contribution bounding.
+        partition_selection: ratio of data dropped because of partition
+          selection.
+
+    """
+    l0: float
+    linf: float
+
+    # This cannot be computed at PartitionSelectionUtility and needs to be
+    # computed for each aggregation separately, since it takes into account data
+    # drop from contribution bounding and that is aggregation-specific.
+    partition_selection: float
+
+
+@dataclass
+class MetricUtility:
+    """Stores aggregate cross-partition metrics for utility analysis.
+
+    Attributes:
+        metric: DP metric for which this analysis was performed.
+        num_dataset_partitions: the number of partitions in dataset.
+        num_non_public_partitions: the number of partitions dropped because
+          of public partitions.
+        num_empty_partitions: the number of empty partitions added because of
+          public partitions.
+        noise_std: the standard deviation of added noise.
+        noise_kind: the noise kind (Laplace or Gaussian)
+        ratio_data_dropped: the information about dropped data.
+        absolute_error: error of (dp_value - actual_value).
+        relative_error: error of (dp_value - actual_value)/actual_value.
+    """
+    metric: pipeline_dp.Metrics
+    num_dataset_partitions: int
+    num_non_public_partitions: int
+    num_empty_partitions: int
+
+    # Noise information.
+    noise_std: float
+    noise_kind: pipeline_dp.NoiseKind
+
+    # Dropped data breakdown.
+    ratio_data_dropped: DroppedDataInfo
+
+    # Value errors
+    absolute_error: ValueErrors
+    relative_error: ValueErrors
+
+
+@dataclass
+class PartitionSelectionUtility:
+    """Stores aggregate metrics about partition selection."""
+
+    strategy: pipeline_dp.PartitionSelectionStrategy
+    num_partitions: float
+    dropped_partitions: MeanVariance
+    ratio_dropped_data: float
+
+
+@dataclass
+class UtilityAnalysisResult:
+    """Stores result of the utility analysis for specific input parameters.
+
+    Attributes:
+        input_aggregate_params: input parameters for which this utility analysis
+          was computed.
+        metric_errors: utility analysis of metrics (e.g. COUNT, SUM, P
```

Comment:
The line break here seems off, can't `P` and `RIVACY_ID_COUNT` be on the same line?